Automattic/jetpack
# Protect: Add ability to auto-fix threats

This PR adds the ability for paid Protect users to trigger auto-fixers for applicable threats.

The "Auto fix" button can be found in three locations:
1. In the details of each individual threat with a fixer available.
2. "Auto-fix all" button above the threat list.
3. "Auto-fix all" button in the admin screen `<Summary />` when there are no threats.

#### Changes proposed in this Pull Request:

* This adds the modals to fix threats. It also adds a modal that redirects to Calypso for the user to add credentials if one is not present.
* Adds a `<CredentialsGate>` component that only renders its children when credentials are available. This gate is implemented in the `<FixThreatModal>` and `<FixAllThreatsModal>` components.
* Adds `fixThreats()`, `fixThreatsStatus()`, and `checkCredentialsState()` redux actions. The first to trigger fixers, and the second to fetch and poll for the fixer status. `fixThreatStatus()` does not resolve until there are no threats with an `in-progress` status.
* Adds three new API endpoints: 
  * `/fix-threats`
  * `/fix-threats-status`
  * `/check-credentials`
* Adds a new `Rewind` class for fetching and working with rewind state.

#### Other information:


#### Jetpack product discussion



https://github.com/Automattic/jetpack/pull/26686

#### Does this pull request change what data or activity we track or use?



No

#### Testing instructions:






* Create a test site with a scan subscription
* Add a bunch of fixable threats to your site
* Open Protect with this branch applied and verify the threats appear.
* Verify that the credentials modal appear when trying to fix a threat before adding the credentials
* Test out the singular fix button in the threat details
* Test out the fix all button

#### Screenshots

<img width="1162" alt="Screen Shot 2022-10-17 at 6 17 14 PM" src="https://user-images.githubusercontent.com/10933065/196307414-afa7332c-0251-405e-ae98-7f97c84dfe44.png">

<img width="658" alt="Screen Shot 2022-10-17 at 6 17 24 PM" src="https://user-images.githubusercontent.com/10933065/196307424-0578ff15-e18f-4366-bd9b-1c57b3164a82.png">

<img width="1116" alt="Screen Shot 2022-10-17 at 6 17 33 PM" src="https://user-images.githubusercontent.com/10933065/196307435-68cc46ef-c982-4dc8-a62e-c077ce2f032e.png">

<img width="1120" alt="Screen Shot 2022-10-17 at 6 24 34 PM" src="https://user-images.githubusercontent.com/10933065/196307446-3afed7aa-8a47-4c5b-9168-eb83908a2d99.png">
